124 doe applications

David-Morales-Norato opened the pull request.

Enmartz commented on colibri/optics/sota_does.py:
```diff
@@ -0,0 +1,90 @@
+    (2017). Design and fabrication of diffractive optical elements with MATLAB.
+    Creates a fresnel lens 
+    Args:
+        N: Number of pixels
```
N is a nonexistent variable, It must be split into `nx` and `ny`, respectively.

David-Morales-Norato replied:
solved

Enmartz commented on colibri/optics/sota_does.py:
```diff
@@ -0,0 +1,90 @@
+
+
+
+def fresnel_lens(ny: int, nx: int, focal = None, radius = None):
```
This code allows the creation of non-square Fresnel lenses. I think all optics should only work with square dimensions to maintain realism.

David-Morales-Norato replied:
I disagree with the restriction to square optics. (That is not necessary real).

Enmartz commented on colibri/optics/functional.py:
```diff
@@ -194,4 +194,563 @@ def backward_spc(y, H):
+def get_space_coords(ny: int, nx: int, pixel_size: float, device=torch.device('cpu'), type='cartesian'):
+    r"""
+    
+    Generate spatial coordinates for wave optics simulations in either Cartesian or polar format.
```
This description should be: Generate coordinates for wave optics simulations in the Cartesian or polar coordinate system.

David-Morales-Norato replied:
solved

Enmartz commented on colibri/optics/functional.py:
```diff
@@ -194,4 +194,563 @@ def backward_spc(y, H):
+    Args:
+        ny (int): Resolution at Y axis in pixels.
+        nx (int): Resolution at X axis in pixels.
+        pixel_size (float): Pixel pixel_size in meters.
```
pixel_size (float): Pixel size in meters.

David-Morales-Norato replied:
solved

Enmartz commented on colibri/optics/sota_does.py:
```diff
@@ -0,0 +1,90 @@
+    return val
+
+
+def spiral_doe(ny: int, nx: int, number_spirals: int, radius: float, focal: float, start_w = 450e-9, end_w = 650e-9):
```
It would be good to write a simple mathematical description of this function if possible.

Enmartz commented on colibri/optics/sota_does.py:
If possible, it would be good to write a simple mathematical description of all functions.

Enmartz commented on colibri/optics/sota_does.py:
```diff
@@ -0,0 +1,90 @@
+    pixel_size = (2*radius)/np.min([ny, nx])
+    r, _ = get_space_coords(ny = ny, nx = nx, pixel_size = pixel_size, type='polar')
+    aperture = circular_aperture(ny = ny, nx = nx, radius = radius, pixel_size = pixel_size)
+    height_map = -(r**2)/(focal)
```
Theoretical question: why is the height map calculated like this? It would be nice if this could be briefly explained in the description of this feature.

David-Morales-Norato replied:
That is the traditional equation of a conventional lens. The equation is a spherical curvature of (focal) $f$

Enmartz commented on examples/demo_does.py:
```diff
@@ -0,0 +1,216 @@
+
+
+fig, ax = plt.subplots(1, 2, figsize=(10, 10))
+ax[0].imshow(height_map, cmap="viridis")
+ax[1].imshow(aperture, cmap="plasma")
```
These images should have a title.

David-Morales-Norato replied:
solved

Enmartz commented on colibri/optics/functional.py:
```diff
@@ -194,4 +194,563 @@ def backward_spc(y, H):
+    .. math::
+        k = \frac{2 \pi}{\lambda}
+    
+    where :math:`\lambda` is the wavelength.
```
This is not well compiled in the documentation, please check.

David-Morales-Norato replied:
solved

Enmartz commented on colibri/optics/functional.py:
```diff
@@ -194,4 +194,563 @@ def backward_spc(y, H):
+    For more information see Goodman, J. W. (2005). Introduction to Fourier optics. Roberts and Company Publishers.
+
+    Args:
+        field (torch.Tensor): Input optical field. Shape (len(wavelengths), nu, nv).
```
`nu` and `nv` are not defined

David-Morales-Norato replied:
solved

Enmartz commented on colibri/optics/functional.py:
```diff
@@ -194,4 +194,563 @@ def backward_spc(y, H):
+
+
+    Args:
+        nu (int): Resolution at X axis in pixels.
+        nv (int): Resolution at Y axis in pixels.
```
There is an ambiguity between the use of `nx` and `ny` with `nu` and `nv`. It should be clarified.

Enmartz replied:
For this and the function `transfer_function_angular_spectrum`

Enmartz commented on colibri/optics/functional.py:
There is an ambiguity between the use of `nx` and `ny` with `nu` and `nv`. It should be clarified.

David-Morales-Norato replied:
You are right, besides it is not consistent with the chosen notation of $(B, C, M, N)$ sizes of spectral images. Therefore we have changed these sizes.

Enmartz commented on colibri/optics/functional.py:
```diff
@@ -194,4 +194,563 @@ def backward_spc(y, H):
+    r"""
+    Apply Fourier convolution theorem to simulate the effect of a linear system characterized by a point spread function (PSF).
+
+    .. math::
+    g = \mathcal{F}^{-1}(\mathcal{F}(f) * \mathcal{F}(h))
+
+    where :math:`f` is the input image, :math:`h` is the PSF, :math:`g` is the convolved output, :math:`\mathcal{F}` and :math:`\mathcal{F}^{-1}` denote the Fourier and inverse Fourier transforms.
```
These equations were not well compiled.

Enmartz commented on colibri/optics/functional.py:
```diff
@@ -194,4 +194,563 @@ def backward_spc(y, H):
+    r"""
+    Apply Fourier convolution theorem to simulate the effect of a linear system characterized by a point spread function (PSF).
+
+    .. math::
+    g = \mathcal{F}^{-1}(\mathcal{F}(f) * \mathcal{F}(h))
```
These equations were not well compiled.

David-Morales-Norato replied:
solved

Enmartz commented on colibri/optics/functional.py:
```diff
@@ -194,4 +194,563 @@ def backward_spc(y, H):
+
+def signal_conv(image: torch.Tensor, psf: torch.Tensor):
+    r"""
+    This function applies the convolution of an image with a Point Spread Function (PSF).
```
This description was not well compiled.

David-Morales-Norato replied:
solved

Enmartz commented on colibri/optics/doe.py:
```diff
@@ -0,0 +1,142 @@
+from colibri.optics.sota_does import fresnel_lens, nbk7_refractive_index
+from .utils import BaseOpticsLayer
+
+class SingleDOESpectral(BaseOpticsLayer):
```
I think this layer should be called SingleDOE, without Spectral, as a more general layer.

David-Morales-Norato replied:
The problem is that it is not general. It will only work for DOEs for spectral applications. 

For Depth and other applications, the math and code is different.

Enmartz commented on colibri/optics/doe.py:
```diff
@@ -0,0 +1,142 @@
+
+    This optical system allow for the capture of spatio-spectral information through a wavelength dependent phase coding of light through a diffractive optical element. 
+    
+    Mathematically, this systems can be described as follows. 
```
this system*

Enmartz replied:
follows: *

David-Morales-Norato replied:
solved

Enmartz commented on colibri/optics/doe.py:
```diff
@@ -0,0 +1,142 @@
+    """
+    def __init__(self, input_shape, 
+                        height_map, 
+                        aperture, 
+                        wavelengths, 
+                        source_distance, 
+                        sensor_distance, 
+                        pixel_size,
```
These parameters should be optional so that by default a simple representation is generated, which is described in the documentation. Just as it was done for the `height_map`. The main idea is to make it easier for the user to implement this layer so that they can use it without having problems.

David-Morales-Norato replied:
solved

Enmartz commented on colibri/optics/doe.py:
```diff
@@ -0,0 +1,142 @@
+                                        approximation=self.approximation)
+    
+    
+    def forward_convolution(self, x, height_map):
```
This function is redundant with `forward` function.

David-Morales-Norato replied:
This function is not redundant because the forward has the `type_calculation` argument used to choose forward_convolution or `deconvolution`. 

But to improve comprehension, I have changed to just `convolution`

Enmartz commented on colibri/optics/functional.py:
```diff
@@ -194,4 +194,563 @@ def backward_spc(y, H):
+    return phase_doe
+
+
+def psf_single_doe_spectral(height_map: torch.Tensor, aperture: torch.Tensor, refractive_index: callable,
```
This function should be called `psf_single_doe`

David-Morales-Norato replied:
Not because only works for spectral

Enmartz commented on colibri/optics/functional.py:
```diff
@@ -194,4 +194,563 @@ def backward_spc(y, H):
+    return psf
+
+
+def addGaussianNoise(y: torch.Tensor, snr: float):
```
Set the name of this function from camel case `addGaussianNoise` to pythonic case `add_gaussian_noise`

David-Morales-Norato replied:
Thanks, solved

Enmartz commented on colibri/optics/functional.py:
```diff
@@ -194,4 +194,563 @@ def backward_spc(y, H):
+    r"""
+    Add Gaussian noise to an image based on a specified signal-to-noise ratio (SNR).
+
+    .. math::
+    y_noisy = y + n
+    .. math::
+    n ~ N(0, \sigma^2)
```
These equations were not well compiled

David-Morales-Norato replied:
solved

Enmartz commented on colibri/optics/functional.py:
```diff
@@ -194,4 +194,563 @@ def backward_spc(y, H):
+    r'''
+    Create a circular aperture mask of a given radius and pixel_size of size (ny, nx).
+
+    .. math::
+        \begin{cases} 
+            1 & \text{if } sqrt(x^2 + y^2) \leq \text{radius} \\
+            0 & \text{otherwise}
+        \end{cases}
```
It is necessary to indicate what the output or operator is. It would also be good to indicate which is the region in which (x, y) operates, whether from [-x, x] or from [0, x_max], the same for y.

David-Morales-Norato replied:
Solved

Enmartz commented on colibri/optics/functional.py:
```diff
@@ -194,4 +194,563 @@ def backward_spc(y, H):
+
+    Simulate the response of an ideal panchromatic sensor by averaging the spectral channels.
+
+    .. math::
+    I = \frac{1}{L} \sum_{\lambda} I_{\lambda}
```
This equation was not well compiled

David-Morales-Norato replied:
solved

Enmartz commented on colibri/optics/functional.py:
```diff
@@ -194,4 +194,563 @@ def backward_spc(y, H):
+
+### Wave optics
+
+def get_space_coords(ny: int, nx: int, pixel_size: float, device=torch.device('cpu'), type='cartesian'):
```
The description should contain a brief mathematical description of what coordinates are generating, and show both cases, "cartesian" and "polar".

David-Morales-Norato replied:
I have improved the documentation of the function. But,  I think this concept is not that hard to understand. That is a math concept of high school.

Enmartz commented on examples/demo_does.py:
```diff
@@ -0,0 +1,216 @@
+manual_device = "cpu"
+doe_size=(100, 100)
+img_size=(200, 200)
+type_doe = "fresnel_lens" # spiral, fresnel_lens
```
I suggest to change this for spiral because it is more visual for this example for the cifar10 dataset

David-Morales-Norato replied:
solved

Enmartz commented on examples/demo_does.py:
```diff
@@ -0,0 +1,216 @@
+img_size=(200, 200)
+type_doe = "fresnel_lens" # spiral, fresnel_lens
+convolution_domain = "fourier" # signal, fourier
+type_wave_propagation = "angular_spectrum" # fresnel, angular_spectrum, fraunhofer
```
Fraunhofer is not implemented, I suggest to remove it from the comment in this line

David-Morales-Norato replied:
Of course it is implemented, see the propagators list

Enmartz replied:
I mean, It is not implementing in the function where it is called. Then, I get the following error when I set  the following line `type_wave_propagation = "fraunhofer" # fresnel, angular_spectrum, fraunhofer` from the `demo_does.py`.

![image](https://github.com/pycolibri/pycolibri/assets/58752635/ba811470-972c-4dbc-b8e4-b3539cc416b8)

David-Morales-Norato replied:
At some point the code to use that case was deleted by mistake. 

I restored the code. 

Solved

Enmartz commented on colibri/optics/functional.py:
```diff
@@ -194,4 +194,563 @@ def backward_spc(y, H):
+        distance (float): Propagation distance in meters.
+        pixel_size (float): Pixel size in meters.
+        wavelength (list): List of wavelengths in meters.
+        approximation (str): Approximation (or diffraction model type) to use, can be "fresnel", "angular_spectrum" or "fraunhofer".
```
Fraunhofer is implemented but it is not being used here

David-Morales-Norato replied:
Because the DOE PSF model needs near-field propagation. Not far field. But we have implemented the Fraunhofer approximation because it was a low effort to future implementations of other systems (depth DOEs, phase retrieval)

Enmartz reviewed (changes requested):
It's a really good improvement for the repository! Thanks for the hard work. As a reviewer, I left you some comments to further improve the documentation and the use of some functions.

David-Morales-Norato commented on colibri/optics/sota_does.py:
```diff
@@ -0,0 +1,90 @@
+
+
+
+def fresnel_lens(ny: int, nx: int, focal = None, radius = None):
```
You are correct. 

The height map is from a conventional lens. Therefore I will change the name to this to improve the comprehension

However, it is important to remark that computationally, both **height maps** are the same **phase** due to the $2\pi$-periodic and wrapping behavior of $\exp(i \cdot \phi_{})$.

David-Morales-Norato commented on colibri/optics/functional.py:
```diff
@@ -194,4 +194,563 @@ def backward_spc(y, H):
+    Simulate the convolutional sensing model of an optical system, using either Fourier or spatial domain methods.
+
+    In the "signal" domain, the convolution operation is performed in the spatial domain using the torch.nn.functional.conv2d function.
+    .. math::
+        g(x, y) = f(x, y) * h(x, y)
+
+    In the "fourier" domain, the convolution operation is performed in the Fourier domain using the Fourier convolution theorem.
+    .. math::
+        \mathcal{G}(f_x, f_y) = \mathcal{F}(f_x, f_y) * \mathcal{H}(f_x, f_y)
```
Solved

David-Morales-Norato commented on colibri/optics/functional.py:
x and y are spatial coordinates, they are not bounded

Enmartz replied:
Yes, but I mean, in terms of implementation, it would be nice if we could tell the user what the operating discrete range is.

David-Morales-Norato replied:
You are right. I have included this description in get  `get_space_coords`

David-Morales-Norato replied:
solved

David-Morales-Norato commented:
> It's a really good improvement for the repository! Thanks for the hard work. As a reviewer, I left you some comments to further improve the documentation and the use of some functions.

Thank you for your exhaustive review, the code and documentation have improved. 

I have addressed all your comments and questions with the necessary changes. Please take a look at my responses and let me know if you have any additional requests or if the PR is ready to be merged.

Enmartz reviewed (changes requested):
I left you some minor comments, please review them, thanks.

David-Morales-Norato commented:
> I left you some minor comments, please review them, thanks.

I have solved your comments

romanjacome99 commented on colibri/optics/functional.py:
```diff
@@ -194,4 +194,563 @@ def backward_spc(y, H):
+
+### Wave optics
+
+def get_space_coords(ny: int, nx: int, pixel_size: float, device=torch.device('cpu'), type='cartesian'):
```
It's a minor thing, but it should be `get_spatial_coords`

romanjacome99 commented on colibri/optics/functional.py:
```diff
@@ -194,4 +194,563 @@ def backward_spc(y, H):
+    r = r.unsqueeze(0)
+    c = torch.exp(1j * wave_number(wavelengths) * distance) / (1j * wavelengths * distance) * torch.exp(1j * wave_number(wavelengths) / (2 * distance) * r**2)
+    c = c.to(device=device)
+    result =  ifft(field / pixel_size**2 / c)
```
according to the equation  $U_0(x,y) \approx  \frac{1}{j\lambda z} e^{j k z} e^{\frac{j k (x^2 + y^2)}{2z}} \mathcal{F}^{-1}\{ U(x,y) \}$ but the code implements $\mathcal{F}^{-1}\left(\frac{U(x,y)}{( \frac{1}{j\lambda z} e^{j k z} e^{\frac{j k (x^2 + y^2)}{2z}}) }\right)$ I don't know if I am confused with the equation.

David-Morales-Norato replied:
The Franhoufer propagation can be described as the operator

$P_{\text{Franhoufer}}: \mathcal{X_0} \rightarrow \mathcal{X_z}$
$U(x, y, z) = P_{\text{Franhoufer}}(U(x, y, 0))$
$U(x,y, z) \approx  \frac{1}{j\lambda z} e^{j k z} e^{\frac{j k (x^2 + y^2)}{2z}} \mathcal{F} \( U(x,y, 0)  \)$

And for some applications we require an inverse propagator  $P_{\text{Franhoufer}}^{-1}: \mathcal{X_z} \rightarrow \mathcal{X_0}$ such that 

$U(x, y, 0) = P_{\text{Franhoufer}}^{-1}(P_{\text{Franhoufer}}(U(x, y, 0)))$


that is $U(x,y, z=0)  \approx \mathcal{F}^{-1}\left(\frac{U(x,y, z)}{( \frac{1}{j\lambda z} e^{j k z} e^{\frac{j k (x^2 + y^2)}{2z}}) }\right)$

The code that you are highlighting is the code of the inverse franhoufer propagation.



pd: For ASM and Fresnel the inversion is just to use (-z) as propagation distance. But in franhouber this is not possible due strong approximations, so we had to implement a different function

romanjacome99 commented on colibri/optics/functional.py:
```diff
@@ -194,4 +194,563 @@ def backward_spc(y, H):
+        img = unpad(img, pad = [0, 0, extra_size[0]//2, extra_size[1]//2])
+    return img
+
+def add_pad(x, pad):
```
This function  and  `unpadd` does not appear in the docs. Please check

David-Morales-Norato replied:
Thanks, solved
